### PR TITLE
`rz_type_db_struct_member_offset()` changes

### PR DESCRIPTION
**Pull request by GustavoLCR**
# PLEASE DO NOT SQUASH 

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

* Rename `rz_type_db_struct_member_offset()`->`rz_type_db_struct_member_packed_offset()`
* Implement `rz_type_db_struct_member_offset()`

**Test plan**

Tests pass

**Closing issues**

None

### CODE REVIEW

File: librz/type/path.c [hunk 1]
```diff
@@ -96,7 +96,7 @@ static st64 path_walker(const RzTypeDB *typedb, const char *path) {
 					return -1;
 				}
 			}
-			offset += rz_type_db_struct_member_offset(typedb, parent->identifier.name, member);
+			offset += rz_type_db_struct_member_packed_offset(typedb, parent->identifier.name, member);
```
Comment (ret2libc) [inline librz/type/path.c:99]: Whether to use packed or not packed shouldn't depend on the struct itself?

**Comment by XVilka** [comment]
This is related to https://github.com/rizinorg/rizin/issues/1241
I personally think this change in the meantime is good enough. We can expand further non-packed structures support at a later stage. @thestr4ng3r your thoughts on this?

**Comment by thestr4ng3r** [comment]
I agree with @ret2libc that the choice of packed/not packed depends on the struct itself, so every struct will have exactly one layout and can never be interpreted in multiple ways.
Thus rizin should either:
* not have an explicit `offset` member at all and consider structs always packed (padding can be done by just inserting stub arrays in between)
* not have an explicit `offset` member at all and have rules attached to a struct that can specify the alignment
* use the explicit `offset` for everything, in which case `rz_type_db_struct_member_packed_offset()` should not exist at all. This has the downside that when a type's size changes, this change might have to be propagated to all structs that contain a member of this type.

If you have a concrete use-case for the current changes in the pr, I am ok-ish with merging it as-is, but this should be clarified asap, otherwise it will be a complete mess again.

**Comment by XVilka** [comment]
@thestr4ng3r it's required for https://github.com/rizinorg/rizin/pull/2177

**Review by thestr4ng3r — APPROVED** [review]

**Comment by ret2libc** [comment]
> I agree with @ret2libc that the choice of packed/not packed depends on the struct itself, so every struct will have exactly one layout and can never be interpreted in multiple ways. Thus rizin should either:
> 
> * not have an explicit `offset` member at all and consider structs always packed (padding can be done by just inserting stub arrays in between)
> * not have an explicit `offset` member at all and have rules attached to a struct that can specify the alignment
> * use the explicit `offset` for everything, in which case `rz_type_db_struct_member_packed_offset()` should not exist at all. This has the downside that when a type's size changes, this change might have to be propagated to all structs that contain a member of this type.
> 
> If you have a concrete use-case for the current changes in the pr, I am ok-ish with merging it as-is, but this should be clarified asap, otherwise it will be a complete mess again.

Do we have the `packed` attribute stored somewhere for a struct BTW?

**Comment by thestr4ng3r** [comment]
No, and that is kind of the point why this needs to be designed right. Just packed/not packed is probably thought too simple for all use-cases.

File: librz/type/path.c [hunk 2]
```diff
@@ -258,3 +258,25 @@ RZ_API ut64 rz_type_db_struct_member_offset(const RzTypeDB *typedb, RZ_NONNULL c
...
 }
+
+/**
+ * \brief Returns the offset in bytes of the structure member if there is a match
+ *
+ * \param typedb Types Database instance
+ * \param name The structure type name
+ * \param name The structure member name
+ */
+RZ_API ut64 rz_type_db_struct_member_offset(const RzTypeDB *typedb, RZ_NONNULL const char *name, RZ_NONNULL const char *member) {
```
Comment (wargio) [inline librz/type/path.c]:
```suggestion
RZ_API ut64 rz_type_db_struct_member_offset(RZ_NONNULL const RzTypeDB *typedb, RZ_NONNULL const char *name, RZ_NONNULL const char *member) {
```

File: librz/type/path.c [hunk 3]
```diff
@@ -235,13 +235,13 @@ RZ_API RZ_OWN RzList /* RzTypePath */ *rz_type_db_get_by_offset(const RzTypeDB *
...
 /**
- * \brief Returns the offset of the structure member if there is a match
+ * \brief Returns the packed offset in bits of the structure member if there is a match
  *
  * \param typedb Types Database instance
  * \param name The structure type name
  * \param name The structure member name
  */
-RZ_API ut64 rz_type_db_struct_member_offset(const RzTypeDB *typedb, RZ_NONNULL const char *name, RZ_NONNULL const char *member) {
+RZ_API ut64 rz_type_db_struct_member_packed_offset(const RzTypeDB *typedb, RZ_NONNULL const char *name, RZ_NONNULL const char *member) {
```
Comment (wargio) [inline librz/type/path.c]:
```suggestion
RZ_API ut64 rz_type_db_struct_member_packed_offset(RZ_NONNULL const RzTypeDB *typedb, RZ_NONNULL const char *name, RZ_NONNULL const char *member) {
```

File: librz/include/rz_type.h [hunk 4]
```diff
@@ -301,6 +301,7 @@ RZ_API void rz_type_path_free(RZ_NULLABLE RzTypePath *tpath);
 RZ_API st64 rz_type_offset_by_path(const RzTypeDB *typedb, RZ_NONNULL const char *path);
 RZ_API RZ_OWN RzList /* RzTypePath */ *rz_type_path_by_offset(const RzTypeDB *typedb, RzBaseType *btype, ut64 offset);
 RZ_API RZ_OWN RzList /* RzTypePath */ *rz_type_db_get_by_offset(const RzTypeDB *typedb, ut64 offset);
+RZ_API ut64 rz_type_db_struct_member_packed_offset(const RzTypeDB *typedb, RZ_NONNULL const char *name, RZ_NONNULL const char *member);
 RZ_API ut64 rz_type_db_struct_member_offset(const RzTypeDB *typedb, RZ_NONNULL const char *name, RZ_NONNULL const char *member);
```
Comment (wargio) [inline librz/include/rz_type.h]:
```suggestion
RZ_API ut64 rz_type_db_struct_member_packed_offset(RZ_NONNULL const RzTypeDB *typedb, RZ_NONNULL const char *name, RZ_NONNULL const char *member);
RZ_API ut64 rz_type_db_struct_member_offset(RZ_NONNULL const RzTypeDB *typedb, RZ_NONNULL const char *name, RZ_NONNULL const char *member);
```

**Comment by XVilka** [comment]
@GustavoLCR please address @wargio's feedback (and rebase on top of the latest `dev`) then we will merge.

**Review by XVilka — APPROVED** [review]

**Pull request merged**